libconcurrent/primitives/threadtools.c

```diff
@@ -59,33 +77,79 @@ inline static void *kthreadWrapper(void *arg) {
-    int ncpus = numa_num_configured_cpus();
-    int nodes = numa_num_task_nodes();
-    int node_size = ncpus / nodes;
-
-    if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) {
-        int half_node_size = node_size / 2;
-        int offset = 0;
-        uint32_t half_cpu_id = pid;
-
-        if (pid >= ncpus / 2) {
-            half_cpu_id = pid - ncpus / 2;
-            offset = ncpus / 2;
+        uint32_t ncpus = numa_num_configured_cpus();
+        uint32_t nodes = numa_num_task_nodes();
+        uint32_t node_size = ncpus / nodes;
+
+        if (__schedule_policy == SYNCH_THREAD_PLACEMENT_NUMA_SPARSE) {
+            if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) { // SMT or HyperThreading detected
+                uint32_t half_node_size = node_size / 2;
+                uint32_t offset = 0;
+                uint32_t half_cpu_id = pid;
+
+                if (pid >= ncpus / 2) {
+                    half_cpu_id = pid - ncpus / 2;
+                    offset = ncpus / 2;
+                }
+                preferred_core = (half_cpu_id % nodes) * half_node_size + half_cpu_id / nodes;
+                preferred_core += offset;
+            } else preferred_core = (pid / nodes) + (pid % nodes) * (node_size);
+        } else if (__schedule_policy == SYNCH_THREAD_PLACEMENT_NUMA_SPARSE_SMT_PREFER) {
+            if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) { // SMT or HyperThreading detected
+                uint32_t double_nodes = 2 * nodes;
+                uint32_t half_node_size = node_size / 2;
+
+                preferred_core = (pid % node_size) * half_node_size + (pid / double_nodes);
+            } else preferred_core = (pid / nodes) + (pid % nodes) * (node_size);
+        } else if (__schedule_policy == SYNCH_THREAD_PLACEMENT_NUMA_DENSE) {
+            preferred_core = pid;
+        } else if (__schedule_policy == SYNCH_THREAD_PLACEMENT_NUMA_DENSE_SMT_PREFER){
+            preferred_core = (pid / nodes) + (pid % nodes) * (node_size);
+        } else {
+            fprintf(stderr, "ERROR: Unsupported scheduling policy: 0x%X\n", __schedule_policy);
+            preferred_core = (pid / nodes) + (pid % nodes) * (node_size);
         }
-        prefered_core = (half_cpu_id % nodes) * half_node_size + half_cpu_id / nodes;
-        prefered_core += offset;
+#else
+        preferred_core = pid;
+#endif
+    }
+    preferred_core %= synchGetNCores();
+
+    return preferred_core;
+}
+
+inline uint32_t synchPreferredNumaNodeOfThread(uint32_t pid) {
+    uint32_t preferred_node = 0;
+
+#ifdef SYNCH_NUMA_SUPPORT
+    uint32_t preferred_core = synchPreferredCoreOfThread(pid);
+    uint32_t ncpus = numa_num_configured_cpus();
+    uint32_t nodes = numa_num_task_nodes();
+    uint32_t node_size = ncpus / nodes;
+
+    if (numa_node_of_cpu(0) == numa_node_of_cpu(ncpus / 2)) { // SMT or HyperThreading detected
+        uint32_t half_node_size = node_size / 2;
+        uint32_t ncpus = numa_num_configured_cpus();
```

Comment:
Already defined in line 133

Reply:
Cleaned up. Check c3c8c996cfa5287a9136c5fb67cb0f8b17ed71aa.